content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+| **Scoped Enums** | [`BHttpStatusCode`](https://git.haiku-os.org/haiku/tree/headers/private/netservices2/HttpResult.h?h=r1beta4#n40) | The API uses [`enum class`](https://en.cppreference.com/w/cpp/language/enum) where it makes sense. |
+| **Lambdas** | [`BHttpSession::Request::SendMessage()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpSession.cpp?h=r1beta4#n1052) | [C++ lambda's](https://www.sandordargo.com/blog/2018/12/19/c++-lambda-expressions) allows the definition of anonymous functions. In the example, this is used by a the `SendMessage()` method to allow the caller to add additional information to the `BMessage` that will be sent to an observer/listener. The alternative would be, that the caller of the method would create a BMessage, which would then be enhanced by the `SendMessage()` method. That alternative does pose a dilemma about safe ownership, or copying. This solution circumvents that. |
+
+In addition to these features, the implementation of the kit also uses some other modern language features internally, like [initializer lists](https://en.cppreference.com/w/cpp/language/constructor), the `auto` keyword for variable declaration, `constexpr`, and [string literals](https://en.cppreference.com/w/cpp/language/string_literal).
+
+### Exceptions for Error Handling
+
+The new HTTP Network Services Kit uses exceptions for error handling, as an alternative to the traditional way of error handling in the APIs inherited from BeOS. In the current public API, error codes are the way to identify and handle errors: functions and methods that may fail have a `status_t` return type, as well as the use of negative values in `ssize_t` or `off_t` return values to signal errors. For many objects, there is an `InitCheck()` method that should be called in order to check if the object is valid and working correctly. For consistency reasons, all new public APIs and most other code do not use exceptions. There is some precedent for experimentation with exceptions, for example in the [private parts of the package kit](https://git.haiku-os.org/haiku/tree/src/kits/package/hpkg?h=r1beta4).
+
+The preview of the HTTP Network Kit requires the user of the API to do all error handling through exceptions. It has the following characteristics:
+* All error handling is done through exceptions. This means that no method will return `status_t` or overload any other return value to signal an error has occured. It will no longer be necessary to call `InitCheck()` after creating a new object, any issues in the constructor will raise an exception.
```

Comment:
```suggestion
* All error handling is done through exceptions. This means that no method will return `status_t` or overload any other return value to signal an error has occurred. It will no longer be necessary to call `InitCheck()` after creating a new object, any issues in the constructor will raise an exception.
```